### DISCUSSION TITLE
How can I increase speed when using multiple smartphones?

### DISCUSSION THREAD

**Discussion by myan-o** [ggml-org/llama.cpp#8500] [Q&A]
I tried using the RPC function, but it ended up being slower. Is there something wrong with the split approach?

**Comment by ggerganov** [comment]
The RPC backend cannot increase the speed - it only distributes the model across multiple devices but evaluates it sequentially

**Reply by myan-o** [reply]
Understood. So this is an effective solution when memory is low. In theory, if the input has already been processed, you can run nodes in parallel, right?

**Reply by ggerganov** [reply]
Hm, not sure what you mean - maybe illustrate with an example what you are thinking of.

The RPC backend currently does not support neither Pipeline nor Tensor parallelism, so all nodes compute sequentially

**Reply by myan-o** [reply]
example

<layer1(input:input_ids)>
-node1(input:input_ids,output:node1_out)
<layer2(input:node1_out)>
-node2(input:node1_out,output:node2_out)
-node3(input:node1_out,output:node3_out)
-node4(input:node1_out,output:node4_out)

If the node1_out variable a has already been processed, node2/node3/node4 can be processed simultaneously, right?

**Reply by myan-o** [reply]
Can parts that are executed in parallel using openmp be distributed using RPC?

**Reply by ggerganov** [reply]
> example

Seems like you are referring to [Tensor Parallelism](https://huggingface.co/docs/text-generation-inference/en/conceptual/tensor_parallelism) over the network. This is not supported currently

> Can parts that are executed in parallel using openmp be distributed using RPC?

It has to be implemented at a higher level

**Reply by myan-o** [reply]
I think it would be easy to implement if you know the code in llama.cpp, but I don't know it, so it's going to be a pain.

**Reply by myan-o** [reply]
Would it be possible to split cblas_sgemm in ggml_blas.c and perform distributed computing?
